source/blog/2022-11-17-announcing-hanami-200rc.html.markdown

```diff
@@ -0,0 +1,357 @@
+
+From routes we move to actions, the classes for handling individual HTTP endpoints. In Hanami 2.0 we’ve redesigned actions to fit seamlessly with the rest of your app.
+
+In your actions you can now `include Deps` like any other class, which makes it to keep your business logic separate and your actions focused on HTTP interactions only:
+
+```ruby
+module Bookshelf
+  module Actions
+    module Books
+      class Show < Bookshelf::Action
+        include Deps["book_repo"]
+
+        params do
+          required(:slug).value(:string)
+        end
+
+        def handle(request, response)
+          book = book_repo.find_by_slug(request.params[:slug])
+
+          response.body = book.to_json
+        end
+      end
+    end
+  end
+end
+```
+
+Actions in Hanami still provide everything you need for a full-featured HTTP layer, including built-in parameter validation, Rack integration, session and cookie handling, flash messages, before/after callbacks and more.
+
+### Type-safe app settings
+
+For your app to do its thing, you'll want to give it various settings: behavioral toggles, API keys, external system URLs and the like. Hanami 2.0 provides a built-in settings class along with [dotenv][dotenv] integration for loading these settings from `.env*` files in local development.
+
+You can define your settings along with type constructors in `config/settings.rb`:
+
+```ruby
+module Bookshelf
+  class Settings < Hanami::Settings
+    setting :email_client_api_key, constructor: Types::String
+    setting :emails_enabled, constructor: Types::Params::Bool
+  end
+end
+```
+
+Your settings are then available as a `"settings"` component for you to use from anywhere in your app:
+
+```ruby
+# Given ENV["EMAILS_ENABLED"] = "true"
+Hanami.app["settings"].emails_enabled # => true
+```
+
+This means you can include `"settings"` in a list of `Deps` and access your settings from any class:
+
+```ruby
+module Bookshelf
+  module Emails
+    class DailyUpdate
+      include Deps["email_client", "settings"]
+
+      def deliver(recipient)
+        return unless settings.emails_enabled
+        # ...
+      end
+    end
+  end
+end
+```
+
+Hanami also loads your settings early in app boot and will raise an error for any settings that fail to meet their type expectations, giving you early feedback and ensuring your app doesn’t boot in a potentially invalid state.
+
+[dotenv]: https://github.com/bkeepers/dotenv
+
+### Providers
+
+So far we’ve seen how your classes and settings can be accessed as components in your app. But what about components that require special handling to initialize? For these, Hanami 2.0 introduces providers. Here’s one for the `"email_client"` component we used earlier, in `config/providers/email_client.rb`:
+
+```ruby
+Hanami.app.register_provider :email_client do
+  prepare do
+    require "acme_email/client"
+  end
+
+  start do
+    client = AcmeEmail::Client.new(
+      api_key: target["settings"].email_client_api_key,
```

Comment:
I reckon `target` may be confusing here because nothing mentions "target" before this code example. I wish we had `Provider#settings` shortcut, actually, something to consider in post-2.0 world.

Anyhow, could we maybe just mention before the code example that providers can **register** components in the **target** container, which is by default your application's container?

Reply:
Rather than trying too hard to introduce all those terms in the text (which I'm trying to keep as light and easy-flowing as possible), here's what I've done...

First, mention that the job of providers is to register. First sentence in the preceding paragraph changed to:

> So far we’ve seen how your classes and settings can be accessed as components in your app. But what about components that require special handling before they’re registered?

Then in the code, I extracted settings assignment into a separate line and preceded it with a comment:

```ruby
    # Grab our settings component from the app (the "target" for this provider)
    settings = target["settings"]
```

Hopefully this should be enough for the code example to be self-explanatory enough. Given we introduced the settings in the section just above, and we're using the same `email_client_api_key` from that example, I think this should hopefully be enough.